### PR TITLE
fix to reduce test flakiness

### PR DESCRIPTION
**Pull request by ryansteakley**
Adds a debug logging statement to see status when status does exist but arn does not. Additionally adds a random sleep in the `create_sagemaker_resource` method. This is because there may be many of the same resource created at the same time and it will take longer than 30 seconds for the `wait_resource_consumed_by_controller `to pass instead of adding extra time. Eventually they will all create but for the tests they will fail at the assert since it only waits 30 seconds.

### CODE REVIEW

File: test/e2e/common/fixtures.py [hunk 1]
```diff
@@ -51,6 +59,7 @@ def xgboost_churn_endpoint(sagemaker_client):
         replacements=replacements,
     )
     assert model_resource is not None
+    sleep_until_k8s_resource_arn(model_resource)
```
Comment (surajkota) [inline test/e2e/common/fixtures.py:62]: remove sleep_until_k8s_resource_arn method and please make a PR to accept wait_period and period length in [load_and_create_resource](https://github.com/aws-controllers-k8s/test-infra/blob/53f590e6a5831ec32824d41de2ca88e9b5c22135/src/acktest/k8s/resource.py#L94) method of acktest package

wait is already part of [wait_resource_consumed_by_controller](https://github.com/aws-controllers-k8s/test-infra/blob/53f590e6a5831ec32824d41de2ca88e9b5c22135/src/acktest/k8s/resource.py#L213) method which can be re-used

there shouldn't be any additional wait needed for arn since it's returned onCreate and wait_resource_consumed_by_controller method is checking if `status` != nil. If status is not nil, arn should be there

File: test/e2e/common/fixtures.py [hunk 2]
```diff
@@ -100,6 +111,7 @@ def xgboost_churn_data_quality_job_definition(xgboost_churn_endpoint):
         spec_file="data_quality_job_definition_xgboost_churn",
         replacements=replacements,
     )
+    sleep(cfg.SLEEP_TIME)
```
Comment (surajkota) [inline test/e2e/common/fixtures.py:114]: same comment as above

File: test/e2e/tests/test_adopt_endpoint.py [hunk 3]
```diff
@@ -203,12 +203,14 @@ def test_smoke(self, sdk_endpoint, adopted_endpoint):
             adopt_config_reference,
             adopt_endpoint_reference,
         ):
-            assert k8s.wait_on_condition(reference, "ACK.Adopted", "True")
+            assert k8s.wait_on_condition(reference, "ACK.Adopted", "True", 12, 60)
```
Comment (surajkota) [inline test/e2e/tests/test_adopt_endpoint.py:206]: shouldn't take 12mins, there is some other problem. Debug this separately

File: test/e2e/tests/test_adopt_endpoint.py [hunk 4]
```diff
@@ -38,7 +38,7 @@
 
 @pytest.fixture(scope="module")
 def name_suffix():
-    return random_suffix_name("sdk-endpoint", 32)
+    return random_suffix_name("sdk-endpoint", 48)
```
Comment (surajkota) [inline test/e2e/tests/test_adopt_endpoint.py:41]: curious: why is this needed?

File: test/e2e/tests/test_endpoint.py [hunk 5]
```diff
@@ -224,12 +236,12 @@ def create_endpoint_test(self, sagemaker_client, xgboost_endpoint):
         assert_endpoint_status_in_sync(
             endpoint_name, reference, cfg.ENDPOINT_STATUS_CREATING
         )
-        assert k8s.wait_on_condition(reference, "ACK.ResourceSynced", "False")
+        assert k8s.wait_on_condition(reference, "ACK.ResourceSynced", "False", 3, 60)
```
Comment (surajkota) [inline test/e2e/tests/test_endpoint.py:239]: are we seeing this assert fail or just increasing timeout for everything?

this was intentional. 3 min is too long for resource to be in sync with service it should happen in 30-1min since that's the requeue time and the check above is checking the status on the server side

same is valid for all `ACK.ResourceSynced` condition waits

File: test/e2e/tests/test_endpoint.py [hunk 6]
```diff
@@ -279,10 +290,10 @@ def update_endpoint_failed_test(
         current_config_name = endpoint_resource["status"].get(
             "latestEndpointConfigName"
         )
-        assert (
-            current_config_name is not None
-            and current_config_name
-            == old_config_resource["spec"].get("endpointConfigName", None)
+        assert current_config_name is not None and current_config_name == old_config_resource[
```
Comment (surajkota) [inline test/e2e/tests/test_endpoint.py:293]: why these formatting changes, are you using a different tool/setting ?

please use black

File: test/e2e/tests/test_hpo.py [hunk 7]
```diff
@@ -132,7 +140,7 @@ def test_stopped(self, xgboost_hpojob):
             k8s.get_resource_arn(resource) == hpo_sm_desc["HyperParameterTuningJobArn"]
         )
         assert hpo_sm_desc["HyperParameterTuningJobStatus"] == cfg.JOB_STATUS_INPROGRESS
-        assert k8s.wait_on_condition(reference, "ACK.ResourceSynced", "False")
+        assert k8s.wait_on_condition(reference, "ACK.ResourceSynced", "False", 3, 60)
```
Comment (surajkota) [inline test/e2e/tests/test_hpo.py:143]: same as above, only add waits where needed

File: test/e2e/tests/test_monitoring_schedule.py [hunk 8]
```diff
@@ -166,34 +168,34 @@ def test_smoke(
...
         # OR
         # create the schedule with a on-going baseline job where it waits for the baselining job to complete
-        assert (
-            wait_resource_monitoring_schedule_status(
-                reference, self.STATUS_PENDING, 5, 2
-            )
-            == self.STATUS_PENDING
-        )
-        assert k8s.wait_on_condition(reference, "ACK.ResourceSynced", "False", 5, 2)
+        # assert (
```
Comment (surajkota) [inline test/e2e/tests/test_monitoring_schedule.py:171]: it's fine to remove this check if it's flaky. is it flaky?

**Review by surajkota — CHANGES_REQUESTED** [review]

File: test/e2e/common/fixtures.py (hunk 1 above)
Comment (ryansteakley) [reply test/e2e/common/fixtures.py:62]: Am rewriting the PR now

File: test/e2e/tests/test_endpoint.py (hunk 6 above)
Comment (ryansteakley) [reply test/e2e/tests/test_endpoint.py:293]: I am using black not sure why they changed

File: test/e2e/tests/test_monitoring_schedule.py (hunk 8 above)
Comment (ryansteakley) [reply test/e2e/tests/test_monitoring_schedule.py]: was flaky because of some sleeps that were added however they have been removed

File: test/e2e/tests/test_endpoint.py [hunk 9]
```diff
@@ -13,6 +13,7 @@
 """Integration tests for the SageMaker Endpoint API.
 """
 
+from _pytest import config
```
Comment (jaypipes) [inline test/e2e/tests/test_endpoint.py:16]: unused import...?

File: test/e2e/tests/test_adopt_endpoint.py [hunk 10]
```diff
@@ -134,6 +134,9 @@ def adopted_endpoint(sdk_endpoint):
     adopt_model_reference, _, adopt_model_resource = create_adopted_resource(
         replacements=replacements,
     )
+    adopt_model_resource = k8s.wait_resource_consumed_by_controller(
+        adopt_model_reference
+    )
```
Comment (jaypipes) [inline test/e2e/tests/test_adopt_endpoint.py:139]: you're already calling `k8s.wait_resource_consumed_by_controller` in the `test_smoke` below for each of these resources...

File: test/e2e/tests/test_endpoint.py [hunk 11]
```diff
@@ -51,6 +52,7 @@ def single_container_model(name_suffix):
         spec_file="xgboost_model",
         replacements=replacements,
     )
+    model_resource = k8s.wait_resource_consumed_by_controller(model_reference)
```
Comment (jaypipes) [inline test/e2e/tests/test_endpoint.py:55]: same comment as above... you are already calling the equivalent of `k8s.wait_resource_consumed_by_controllers` below for all resources involved in these tests...

**Review by jaypipes — CHANGES_REQUESTED** [review]

**Comment by RedbackThomson** [comment]
Should have fixed the Helm dependency error
/retest

File: test/e2e/tests/test_model_bias_job_definition.py [hunk 12]
```diff
@@ -23,6 +23,7 @@
 )
 from e2e.replacement_values import REPLACEMENT_VALUES
 from e2e.common.fixtures import xgboost_churn_endpoint
+from e2e.common import config as cfg
```
Comment (jaypipes) [inline test/e2e/tests/test_model_bias_job_definition.py:26]: unused import?

File: test/e2e/tests/test_model_explainability_job_definition.py [hunk 13]
```diff
@@ -23,6 +23,7 @@
 )
 from e2e.replacement_values import REPLACEMENT_VALUES
 from e2e.common.fixtures import xgboost_churn_endpoint
+from e2e.common import config as cfg
```
Comment (jaypipes) [inline test/e2e/tests/test_model_explainability_job_definition.py:26]: unused import?

File: test/e2e/tests/test_model_quality_job_definition.py [hunk 14]
```diff
@@ -23,6 +23,7 @@
 )
 from e2e.replacement_values import REPLACEMENT_VALUES
 from e2e.common.fixtures import xgboost_churn_endpoint
+from e2e.common import config as cfg
```
Comment (jaypipes) [inline test/e2e/tests/test_model_quality_job_definition.py:26]: unused import

File: test/e2e/tests/test_monitoring_schedule.py [hunk 15]
```diff
@@ -24,6 +24,7 @@
     xgboost_churn_data_quality_job_definition,
     xgboost_churn_endpoint,
 )
+from e2e.common import config as cfg
```
Comment (jaypipes) [inline test/e2e/tests/test_monitoring_schedule.py:27]: unused import

File: test/e2e/tests/test_processingjob.py [hunk 16]
```diff
@@ -47,6 +46,8 @@ def kmeans_processing_job():
     )
 
     assert resource is not None
+    if k8s.get_resource_arn(resource) is None:
+        print(resource["status"])
```
Comment (jaypipes) [inline test/e2e/tests/test_processingjob.py:50]: please use `logging.debug()` instead of raw print statements

File: test/e2e/tests/test_trainingjob.py [hunk 17]
```diff
@@ -47,6 +46,8 @@ def xgboost_training_job():
     )
 
     assert resource is not None
+    if k8s.get_resource_arn(resource) is None:
+        print(resource["status"])
```
Comment (jaypipes) [inline test/e2e/tests/test_trainingjob.py:50]: `logging.debug`

File: test/e2e/tests/test_trainingjob_debugger.py [hunk 18]
```diff
@@ -45,7 +44,8 @@ def xgboost_training_job_debugger():
         spec_file="xgboost_trainingjob_debugger",
         replacements=replacements,
     )
-
+    if k8s.get_resource_arn(resource) is None:
+        print(resource["status"])
```
Comment (jaypipes) [inline test/e2e/tests/test_trainingjob_debugger.py:48]: `logging.debug`

File: test/e2e/tests/test_transformjob.py [hunk 19]
```diff
@@ -78,8 +79,10 @@ def xgboost_transformjob(xgboost_model_for_transform):
         spec_file="xgboost_transformjob",
         replacements=replacements,
     )
-
+
```
Comment (jaypipes) [inline test/e2e/tests/test_transformjob.py:82]: whitespace above?

File: test/e2e/tests/test_transformjob.py [hunk 20]
```diff
@@ -78,8 +79,10 @@ def xgboost_transformjob(xgboost_model_for_transform):
         spec_file="xgboost_transformjob",
         replacements=replacements,
     )
-
+  
     assert resource is not None
+    if k8s.get_resource_arn(resource) is None:
+        print(resource["status"])
```
Comment (jaypipes) [inline test/e2e/tests/test_transformjob.py:85]: logging.debug

File: test/e2e/tests/test_endpoint.py [hunk 21]
```diff
@@ -52,6 +53,8 @@ def single_container_model(name_suffix):
         replacements=replacements,
     )
     assert model_resource is not None
+    if k8s.get_resource_arn(model_resource) is None:
+        print(model_resource["status"])
```
Comment (jaypipes) [inline test/e2e/tests/test_endpoint.py:57]: here and throughout, please use `logging.debug` instead of raw `print` statements.

File: test/e2e/tests/test_endpoint.py (hunk 11 above)
Comment (jaypipes) [reply test/e2e/tests/test_endpoint.py]: @ryansteakley wanted to follow up on this comment. Please note that when I said you are already calling the equivalent of `k8s.wait_resource_consumed_by_controller`, I was referring to the fact that in test_endpoint.py, you are calling `k8s.wait_on_condition`:

https://github.com/aws-controllers-k8s/sagemaker-controller/blob/17c03a5e60d360240ede43928f645616325bc0b4/test/e2e/tests/test_endpoint.py#L232

after creating the CR in kubernetes. And that is essentially the equivalent of `k8s.wait_resource_consumed_by_controller`

File: test/e2e/common/fixtures.py [hunk 22]
```diff
@@ -64,6 +66,8 @@ def xgboost_churn_endpoint(sagemaker_client):
         replacements=replacements,
     )
     assert endpoint_config_resource is not None
+    if k8s.get_resource_arn(endpoint_config_resource) is None:
+        logging.debug(endpoint_config_resource["status"])
```
Comment (surajkota) [inline test/e2e/common/fixtures.py:70]: nit: add some human friendly stuff like arn is None, status is:

**Review by surajkota — COMMENTED** [review]
Thanks for work on this PR!!
Very important for day to day developer experience

Please squash the commits

File: test/e2e/__init__.py [hunk 23]
```diff
@@ -36,12 +37,19 @@ def sagemaker_client():
...
+    replacements,
+    namespace="default",
+    wait_period=3,
+    period_length=10,
 ):
     """
     Wrapper around k8s.load_and_create_resource to create a SageMaker resource
     """
-
+    rand = random.randrange(1, 4)
```
Comment (surajkota) [inline test/e2e/__init__.py:51]: nit: add a comment about it

Comment (ryansteakley) [reply test/e2e/__init__.py:55]: added

**Review by jaypipes — APPROVED** [review]
:+1:

**Review by surajkota — APPROVED** [review]
/approve
/lgtm
